synapse/federation/sender/per_destination_queue.py

```diff
@@ -580,3 +501,125 @@ def _start_catching_up(self) -> None:
         """
         self._catching_up = True
         self._pending_pdus = []
+
+
+@attr.s(slots=True)
+class _TransactionQueueManager:
+    """A helper async context manager for pulling stuff off the queues and
+    tracking what was last successfully sent, etc.
+    """
+
+    queue = attr.ib(type=PerDestinationQueue)
+
+    _device_stream_id = attr.ib(type=Optional[int], default=None)
+    _device_list_id = attr.ib(type=Optional[int], default=None)
+    _last_stream_ordering = attr.ib(type=Optional[int], default=None)
+
+    async def __aenter__(self) -> Tuple[List[EventBase], List[Edu]]:
+        # We have to keep 2 free slots for presence and rr_edus
+        limit = MAX_EDUS_PER_TRANSACTION - 2
+
+        device_update_edus, dev_list_id = await self.queue._get_device_update_edus(
+            limit
+        )
+
+        if device_update_edus:
+            self._device_list_id = dev_list_id
+        else:
+            self.queue._last_device_list_stream_id = dev_list_id
+
+        limit -= len(device_update_edus)
+
+        (
+            to_device_edus,
+            device_stream_id,
+        ) = await self.queue._get_to_device_message_edus(limit)
+
+        if to_device_edus:
+            self._device_stream_id = device_stream_id
+        else:
+            self.queue._last_device_stream_id = device_stream_id
+
+        pending_edus = device_update_edus + to_device_edus
+
+        pending_pdus = self.queue._pending_pdus
+
+        # We can only include at most 50 PDUs per transactions
+        pending_pdus, self.queue._pending_pdus = pending_pdus[:50], pending_pdus[50:]
+
+        pending_edus.extend(self.queue._get_rr_edus(force_flush=False))
+        pending_presence = self.queue._pending_presence
+        self.queue._pending_presence = {}
```

Comment:
This will cause us to declare a new dict in some cases when we don't need to, it seems we could (instead) check `if self.queue._pending_presence` below and only overwrite this value if necessary (i.e. the `pending_presence` variable is unnecessary).